### PR TITLE
Add new options for distribution build to tag docker with build number

### PR DESCRIPTION
**Pull request by peterzhuamazon**
Signed-off-by: Peter Zhu <zhujiaxi@amazon.com>

### Description
Add new options for distribution build to tag docker with build number

### Issues Resolved
https://github.com/opensearch-project/opensearch-build/issues/1762

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by zelinh — APPROVED** [review]

File: jenkins/docker/docker-build.jenkinsfile [hunk 1]
```diff
@@ -40,18 +35,15 @@ pipeline {
...
-                    }
+                    def CREDENTIAL_ID = "jenkins-staging-dockerhub-credential"
                     sh "echo Account: $CREDENTIAL_ID"
                     withCredentials([usernamePassword(credentialsId: CREDENTIAL_ID, usernameVariable: 'DOCKER_USERNAME', passwordVariable: 'DOCKER_PASSWORD')]) {
                         sh '''
-                            set -ex
+                            set -e
                             echo Login to $CREDENTIAL_ID
-                            docker logout && docker login -u $DOCKER_USERNAME -p $DOCKER_PASSWORD && eval $DOCKER_BUILD_SCRIPT_WITH_COMMANDS
+                            docker logout && echo $DOCKER_PASSWORD | docker login -u $DOCKER_USERNAME --password-stdin && eval $DOCKER_BUILD_SCRIPT_WITH_COMMANDS
```
Comment (gaiksaya) [inline jenkins/docker/docker-build.jenkinsfile:46]: Echoing password is not a good practice. Is this change intended?

File: jenkins/opensearch-dashboards/distribution-build.jenkinsfile [hunk 2]
```diff
@@ -38,10 +38,10 @@ pipeline {
             defaultValue: "bwc-test-opensearch-dashboards",
             trim: true
         )
-        booleanParam(
+        choice(
             name: 'BUILD_DOCKER',
-            description: 'Build docker image or not.',
-            defaultValue: true
+            description: 'Build docker image or not with options.',
+            choices: ['build_docker', 'build_docker_with_build_number_tag', 'do_not_build_docker'],
```
Comment (gaiksaya) [inline jenkins/opensearch-dashboards/distribution-build.jenkinsfile:44]: Wanted to confirm if programmatically triggering this workflow might be affected by this change

**Review by gaiksaya — CHANGES_REQUESTED** [review]

File: jenkins/docker/docker-build.jenkinsfile (hunk 1 above)
Comment (peterzhuamazon) [reply jenkins/docker/docker-build.jenkinsfile:46]: @prudhvigodithi 
This is intended and it is actually much safer than the original setup.
Note that the password is pipe directly into --password-stdin instead of stdout.
In that case `docker login` command will not print `login with -u **** -p ****` which is the best practice from docker.

File: jenkins/opensearch-dashboards/distribution-build.jenkinsfile (hunk 2 above)
Comment (peterzhuamazon) [reply jenkins/opensearch-dashboards/distribution-build.jenkinsfile:44]: I have already taken care of all the instances that get touched by this change.
Thanks.

File: jenkins/docker/docker-build.jenkinsfile (hunk 1 above)
Comment (gaiksaya) [reply jenkins/docker/docker-build.jenkinsfile:46]: But when you do echo $DOCKER_PASSWORD it will emit an password right?

File: jenkins/opensearch-dashboards/distribution-build.jenkinsfile (hunk 2 above)
Comment (peterzhuamazon) [reply jenkins/opensearch-dashboards/distribution-build.jenkinsfile:44]: Jenkins default the 1st entry in the choices list is the default, thus the behavior is exactly the same as before and I have changed all the instance to reflect this change.

Comment (gaiksaya) [reply jenkins/opensearch-dashboards/distribution-build.jenkinsfile:44]: I guess you didn't get my question. I was concerned if choices and parameters have different affect on the workflow. If build-workflow triggers docker-workflow, can we select an option like default choice or something?

File: jenkins/docker/docker-build.jenkinsfile (hunk 1 above)
Comment (peterzhuamazon) [reply jenkins/docker/docker-build.jenkinsfile:46]: It send the password to the pipe it did not emit to anywhere else.

File: jenkins/opensearch-dashboards/distribution-build.jenkinsfile (hunk 2 above)
Comment (gaiksaya) [reply jenkins/opensearch-dashboards/distribution-build.jenkinsfile:44]: Ok got it. This was new to me Thanks!

File: jenkins/docker/docker-build.jenkinsfile (hunk 1 above)
Comment (gaiksaya) [reply jenkins/docker/docker-build.jenkinsfile:46]: Okay still seems risky. Better to monitor and maybe enhance later.

Comment (gaiksaya) [reply jenkins/docker/docker-build.jenkinsfile:46]: Why can't we keep what was before? I am not comfortable with echoing password.

Comment (gaiksaya) [reply jenkins/docker/docker-build.jenkinsfile:46]: [Offline discussion] Looks like previously docker did emit the password. This way it will not. Hence, this is better. Thanks!

Comment (zelinh) [reply jenkins/docker/docker-build.jenkinsfile:46]: I see we also use it here. https://github.com/opensearch-project/opensearch-build/blob/92efa806f8bb3321aa0af40711ba6de3b879b93f/vars/copyContainer.groovy#L16

Comment (peterzhuamazon) [reply jenkins/docker/docker-build.jenkinsfile:46]: Thank you!

**Review by gaiksaya — APPROVED** [review]
Recommend to add a note that choices first option is the default. Easy to miss thing.
Thanks!
